### ISSUE TITLE
[Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Invalid column name 'creator'.

### ISSUE THREAD

**Issue by mrkskwsnck** [frankkoch/moodle-mod_studentquiz#314]
Adding and displaying this activity leads to an SQL Server error. The activity does not work at all.

Maybe the most important fact is that our Moodle platform runs with Microsoft SQL Server as database.

**System**

- Moodle 3.9.4+ (Build: 20210226)
- DB: Microsoft SQL Server
- Plug-in: StudentQuiz v4.4.2 (Build: 2021022500)

**Steps to reproduce**

- Set up Moodle with SQL Server
- Install plug-in StudentQuiz
- Within a course add the activity StudentQuiz
- Click Save and display
- The error message appears immediately

**Screenshot**

![image](https://user-images.githubusercontent.com/486983/110654307-4ec16a80-81be-11eb-8728-d71d74a304a1.png)

**Full error message**

```
Debug info: SQLState: 42S22<br>
Error Code: 207<br>
Message: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Invalid column name 'creator'.<br>
SQLState: 42S22<br>
Error Code: 207<br>
Message: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Invalid column name 'creator'.<br>

SELECT statspercategory.userid AS userid,
-- Aggregate values over all categories in cm context.
-- Note: Max() of equals is faster than Sum() of groups.
-- See: https://dev.mysql.com/doc/refman/5.7/en/group-by-optimization.html.
MAX(points) AS points, MAX(questions_created) AS questions_created,
MAX(questions_created_and_rated) AS questions_created_and_rated, MAX(questions_approved) AS questions_approved,
MAX(questions_disapproved) AS questions_disapproved,
MAX(rates_received) AS rates_received, MAX(rates_average) AS rates_average,
MAX(question_attempts) AS question_attempts, MAX(question_attempts_correct) AS question_attempts_correct,
MAX(question_attempts_incorrect) AS question_attempts_incorrect,
MAX(last_attempt_exists) AS last_attempt_exists, MAX(last_attempt_correct) AS last_attempt_correct,
MAX(last_attempt_incorrect) AS last_attempt_incorrect
-- Select for each question category in context.
FROM (
SELECT u.id AS userid, qc.id AS category,
-- Calculate points.
COALESCE (
ROUND (
-- Questions created.
COALESCE(creators.countq, 0) * '10' +
-- Questions approved.
COALESCE(approvals.approved, 0) * '5' +
-- Rating.
COALESCE(rates.avgv, 0) * (COALESCE(creators.countq, 0) -
COALESCE(rates.not_rated_questions, 0)) * '3' +
-- Correct answers.
COALESCE(lastattempt.last_attempt_correct, 0) * '2' +
-- Incorrect answers.
COALESCE(lastattempt.last_attempt_incorrect, 0) * '0',
1
),
0
) AS points,
-- Questions created.
COALESCE(creators.countq, 0) AS questions_created,
-- Questions created and rated.
COALESCE(COALESCE(creators.countq, 0) - COALESCE(rates.not_rated_questions, 0),
0) AS questions_created_and_rated,
-- Questions approved.
COALESCE(approvals.approved, 0) AS questions_approved,
-- Questions disapproved.
COALESCE(approvals.disapproved, 0) AS questions_disapproved,
-- Questions rating received.
COALESCE(rates.countv, 0) AS rates_received,
COALESCE(rates.avgv, 0) AS rates_average,
-- Question attempts.
COALESCE(attempts.counta, 0) AS question_attempts,
COALESCE(attempts.countright, 0) AS question_attempts_correct,
COALESCE(attempts.countwrong, 0) AS question_attempts_incorrect,
-- Last attempt.
COALESCE(lastattempt.last_attempt_exists, 0) AS last_attempt_exists,
COALESCE(lastattempt.last_attempt_correct, 0) AS last_attempt_correct,
COALESCE(lastattempt.last_attempt_incorrect, 0) AS last_attempt_incorrect
-- WARNING: the trailing ) is intentionally missing, found in mod_studentquiz_user_stats var statsbycat
-- Following newline is intentional because this string is concatenated
FROM mdl_studentquiz sq
-- Get this Studentquiz Question category.
JOIN mdl_context con ON con.instanceid = sq.coursemodule
AND con.contextlevel = 70
JOIN mdl_question_categories qc ON qc.contextid = con.id
-- Only enrolled users.
JOIN mdl_course c ON c.id = sq.course
JOIN mdl_context cctx ON cctx.instanceid = c.id
AND cctx.contextlevel = 50
JOIN mdl_role_assignments ra ON cctx.id = ra.contextid
JOIN mdl_user u ON u.id = ra.userid
-- Question created by user.
LEFT JOIN (
SELECT count(*) AS countq, q.createdby AS creator
FROM mdl_studentquiz sq
JOIN mdl_context con ON con.instanceid = sq.coursemodule
JOIN mdl_question_categories qc ON qc.contextid = con.id
JOIN mdl_question q ON q.category = qc.id
JOIN mdl_studentquiz_question sqq ON q.id = sqq.questionid
WHERE q.hidden = 0
AND sqq.hidden = 0
AND q.parent = 0
AND sq.coursemodule = '599'
GROUP BY creator
) creators ON creators.creator = u.id
-- Approved questions.
LEFT JOIN (
SELECT count(*) AS countq, q.createdby AS creator,
COUNT(CASE WHEN sqq.state = 0 THEN q.id END) as disapproved,
COUNT(CASE WHEN sqq.state = 1 THEN q.id END) as approved
FROM mdl_studentquiz sq
JOIN mdl_context con ON con.instanceid = sq.coursemodule
JOIN mdl_question_categories qc ON qc.contextid = con.id
JOIN mdl_question q ON q.category = qc.id
JOIN mdl_studentquiz_question sqq ON q.id = sqq.questionid
WHERE q.hidden = 0
AND q.parent = 0
AND sqq.hidden = 0
AND sq.coursemodule = '599'
GROUP BY creator
) approvals ON approvals.creator = u.id
-- Average of Average Rating of own questions.
LEFT JOIN (
SELECT createdby, AVG(avg_rate_perq) AS avgv, SUM(num_rate_perq) AS countv,
SUM(question_not_rated) AS not_rated_questions
FROM (
SELECT q.id, q.createdby AS createdby, AVG(sqv.rate) AS avg_rate_perq,
COUNT(sqv.rate) AS num_rate_perq,
MAX(CASE WHEN sqv.id IS NULL THEN 1 ELSE 0 END) AS question_not_rated
FROM mdl_studentquiz sq
JOIN mdl_context con ON con.instanceid = sq.coursemodule
JOIN mdl_question_categories qc ON qc.contextid = con.id
JOIN mdl_question q ON q.category = qc.id
JOIN mdl_studentquiz_question sqq ON q.id = sqq.questionid
LEFT JOIN mdl_studentquiz_rate sqv ON q.id = sqv.questionid
WHERE q.hidden = 0
AND q.parent = 0
AND sqq.hidden = 0
AND sq.coursemodule = '599'
GROUP BY q.id, q.createdby
) avgratingperquestion
GROUP BY createdby
) rates ON rates.createdby = u.id
LEFT JOIN (
SELECT sp.userid, COUNT(*) AS last_attempt_exists, SUM(lastanswercorrect) AS last_attempt_correct,
SUM(1 - lastanswercorrect) AS last_attempt_incorrect
FROM mdl_studentquiz_progress sp
JOIN mdl_studentquiz sq ON sq.id = sp.studentquizid
JOIN mdl_question q ON q.id = sp.questionid
JOIN mdl_studentquiz_question sqq ON sp.questionid = sqq.questionid
WHERE sq.coursemodule = '599'
AND q.hidden = 0
AND sqq.hidden = 0
GROUP BY sp.userid
) lastattempt ON lastattempt.userid = u.id
LEFT JOIN (
SELECT SUM(attempts) AS counta, SUM(correctattempts) AS countright,
SUM(attempts - correctattempts) AS countwrong, sp.userid AS userid
FROM mdl_studentquiz_progress sp
JOIN mdl_studentquiz sq ON sq.id = sp.studentquizid
JOIN mdl_question q ON q.id = sp.questionid
JOIN mdl_studentquiz_question sqq ON sp.questionid = sqq.questionid
WHERE sq.coursemodule = '599'
AND q.hidden = 0
AND sqq.hidden = 0
GROUP BY sp.userid
) attempts ON attempts.userid = u.id
WHERE sq.coursemodule = '599' AND u.id = '57' ) statspercategory GROUP BY userid
[array (
0 => '10',
1 => '5',
2 => '3',
3 => '2',
4 => '0',
5 => '599',
6 => '599',
7 => '599',
8 => '599',
9 => '599',
10 => '599',
11 => '57',
)]
Error code: dmlreadexception
Stack trace:

    line 486 of /lib/dml/moodle_database.php: dml_read_exception thrown
    line 324 of /lib/dml/sqlsrv_native_moodle_database.php: call to moodle_database->query_end()
    line 431 of /lib/dml/sqlsrv_native_moodle_database.php: call to sqlsrv_native_moodle_database->query_end()
    line 900 of /lib/dml/sqlsrv_native_moodle_database.php: call to sqlsrv_native_moodle_database->do_query()
    line 976 of /lib/dml/sqlsrv_native_moodle_database.php: call to sqlsrv_native_moodle_database->get_recordset_sql()
    line 1624 of /lib/dml/moodle_database.php: call to sqlsrv_native_moodle_database->get_records_sql()
    line 800 of /mod/studentquiz/locallib.php: call to moodle_database->get_record_sql()
    line 123 of /mod/studentquiz/reportlib.php: call to mod_studentquiz_user_stats()
    line 93 of /mod/studentquiz/renderer.php: call to mod_studentquiz_report->get_user_stats()
    line 959 of /mod/studentquiz/renderer.php: call to mod_studentquiz_renderer->render_stat_block()
    line 97 of /mod/studentquiz/view.php: call to mod_studentquiz_overview_renderer->add_fake_block()
```

**Comment by dionysius** [comment]
Your assumption is correct. We do not support MSSQL due to this exact reason, but we want to. Be aware that MSSQL is not listed in the supported databases in the Readme.

Although I think it is probably only this exact (quite big) query that causes issues with MSSQL, the others should be fine. I also think it will be impossible to make this query compatible with good performance to all databases, so I am willing to have a query  dedicated to MSSQL with a database type detection condition.

1) If someone can help me translate this query for MSSQL, I can add this as "unofficial support" and will work at that time, but we can never guarantee that we catch errors while developing this plugin further.

2) To officially support MSSQL we need this plugin to be automatically tested against it as well. That requires that our [testing toolchain](https://github.com/moodlehq/moodle-plugin-ci) supports MSSQL.

**Comment by mrkskwsnck** [comment]
It turns out, it was not that hard to make this plug-in work with SQL Server. I started hacking and the found issues can be broken down to two points:

- SQL Server is not able to `GROUP BY` column aliases. Few minor changes had to be made to the big query mentioned.
- SQL Server 2017 and newer provides the `STRING_AGG` function. That way it can be treated like PostgreSQL.

That being said, SQL Server 2017 and newer can be considered as experimentally supported. So, I created the PR https://github.com/frankkoch/moodle-mod_studentquiz/pull/317
